backends/dpdk/dpdkArch.cpp

```diff
@@ -2374,6 +2402,191 @@ const IR::Node* SplitP4TableCommon::postorder(IR::SwitchStatement* statement) {
+                                                           refMap, typeMap, &isConstructedInPlace,
+                                                           meterExternName);
+    int table_size = getTableSize(tbl);
+    auto table_type = ::get(structure->table_type_map, tbl->name.name);
+
+    // Direct Counter and Meter are not supported with Wildcard match tables
+    if (table_type == InternalTableType::WILDCARD && (counterInstance || meterInstance)) {
+        ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET, "Direct counters and direct meters are"
+                " unsupported for wildcard match table %1%", tbl->name);
+        return;
+    }
+    if (table_type == InternalTableType::LEARNER)
+            table_size *= 4;
+
+    auto default_action = tbl->getDefaultAction();
+    if (default_action) {
+        if (auto mc = default_action->to<IR::MethodCallExpression>()) {
+            default_action = mc->method;
+        }
+
+        auto path = default_action->to<IR::PathExpression>();
+        BUG_CHECK(path, "Default action path %s cannot be found", default_action);
+        if (auto defaultActionDecl = refMap->getDeclaration(path->path)->to<IR::P4Action>()) {
+            if (defaultActionDecl->name.originalName != "NoAction") {
+            if (!ifMethodFound(defaultActionDecl, "count", counterExternName)) {
+                if (counterInstance) {
+                    ::error(ErrorType::ERR_EXPECTED, "Expected default action %1% to have 'count'"
+                            " method call for DirectCounter extern instance %2%",
+                            defaultActionDecl->name, *counterInstance->name);
+                    return;
+                }
+            }
+            if (!ifMethodFound(defaultActionDecl, "execute", meterExternName)) {
+                if (meterInstance) {
+                    ::error(ErrorType::ERR_EXPECTED, "Expected default action %1% to have "
+                            "'execute' method call for DirectMeter extern instance %2%",
+                             defaultActionDecl->name, *meterInstance->name);
+                    return;
+                }
+            }
+        }
+      }
+    }
+
+    if (counterInstance && counterInstance->type->name == "DirectCounter") {
+        directMeterCounterSizeMap.emplace(counterExternName, table_size + 1);
+        structure->direct_resource_map.emplace(counterExternName, tbl);
+    }
+    if (meterInstance && meterInstance->type->name == "DirectMeter") {
+        directMeterCounterSizeMap.emplace(meterExternName, table_size + 1);
+        structure->direct_resource_map.emplace(meterExternName, tbl);
+    }
+}
+
+void ValidateDirectCounterMeter::validateMethodInvocation(P4::ExternMethod *a) {
+    if ((a->originalExternType->getName().name == "DirectCounter" &&
+        a->method->getName().name == "count") ||
+        (a->originalExternType->getName().name == "DirectMeter" &&
+        a->method->getName().name == "execute")) {
```

Comment:
what if the method does not have this name?

Reply:
I suspect the answer is "this is implementing the PSA/PNA architecture definitions of DirectCounter and DirectMeter, and the method names are defined in the spec".  So the methods won't have any other names.

Reply:
that's why an error should be given

Reply:
Sorry, I misunderstood the reason for your question.  Carry on :-)

Reply:
Yes. I will add error here